### ISSUE TITLE
Misc issues (map, const generic, array updates, assertions)

### ISSUE THREAD

**Issue by franziskuskiefer** [AeneasVerif/charon#45]
Here are some issues when trying to run charon on Rust code.

map
```rust
pub fn map(x: [i32; 256],) -> [i32; 256] {
    x.map(|v| v)
}
```

```
thread 'rustc' panicked at 'not implemented: impl source: ImplSourceClosureData(closure_def_id=DefId(0:6 ~ simplemap[2625]::add::{closure#0}), substs=[i16, extern "rust-call" fn((i32,)) -> i32, ()], nested=[Obligation(predicate=Binder(ClosureKind(DefId(0:6 ~ simplemap[2625]::add::{closure#0}), [i16, extern "rust-call" fn((i32,)) -> i32, ()], FnMut), []), depth=1)])', /Users/franziskus/.cargo/git/checkouts/hacspec-v2-fad44c11f2620bcb/60ef8f9/frontend/exporter/src/types/traits.rs:428:13
```

const generic
```rust
pub fn array<const LEN: usize>() -> [u8; LEN] {
    [0u8; LEN]
}
```

```
thread 'rustc' panicked at 'internal error: entered unreachable code: ConstGeneric::as_value: Not the proper variant', src/types.rs:404:56
```

array update in loops
```rust
fn cbd(mut prf_input: [u8; 33]) {
    for i in 0..3 {
        prf_input[0] = i;
    }
}
```

```
thread 'rustc' panicked at 'internal error: entered unreachable code: Could not find a clause for parameter:
- target param: @TraitDecl1<Self::IntoIter>
- available clauses:
@TraitDecl0<Self>', src/translate_predicates.rs:599:13
```

assertions
```rust
pub(crate) fn select(lhs: &[u8], rhs: &[u8]) {
    debug_assert_eq!(lhs.len(), rhs.len());
}
```

```
thread 'rustc' panicked at 'Unsupported ADT: core::panicking::AssertKind', src/translate_functions_to_ullbc.rs:924:33
```

**Comment by sonmarcho** [comment]
Sorry for my late answer: I didn't receive a notification for this issue.

Let me comment those:
- `map`: this is a known issue, which comes from the fact that you are using a function pointer like an instance of `FnMut`. I guess you are using the branch `son_traits`, which is work in progress to add support for traits: this feature is the next in my list.
- `array`: I fixed this on the branch `son_traits`
- `cbd`: I stumbled on that bug with one of my own tests (it is fixed)
- `select`: this comes from the handling of ADTs coming from external dependencies. I've been wanting to fix this for a while but have been distracted with other things. I will do a PR for this after I merge `son_traits`. The quick fix for now is to use a different macro. For example, if you expand this expression to `if !(lhs.len() == rhs.len()) { panic!() }` it will work. If this is the only occurrence of this bug in your codebase, I can also provide a temporary (hacky) fix.
